pkg/rate/api_limiter_test.go

```diff
@@ -716,3 +718,91 @@ func (b *ControllerSuite) TestAdjustedParallelRequests(c *check.C) {
 	a.adjustmentFactor = 1.0
 	c.Assert(a.adjustedParallelRequests(), check.Equals, 2)
 }
+
+func (b *ControllerSuite) TestStressRateLimiter(c *check.C) {
+	a := NewAPILimiter("foo", APILimiterParameters{
+		EstimatedProcessingDuration: 5 * time.Millisecond,
+		RateLimit:                   1000.0,
+		ParallelRequests:            50,
+		RateBurst:                   1,
+		MaxWaitDuration:             10 * time.Millisecond,
+		AutoAdjust:                  true,
+	}, nil)
+
+	var (
+		sem                = semaphore.NewWeighted(100)
+		completed, retries int32
+	)
+
+	go func() {
+		for i := 0; i < 1000; i++ {
+			sem.Acquire(context.Background(), 1)
+			go func() {
+				var (
+					err error
+					req LimitedRequest
+				)
+				for req == nil {
+					req, err = a.Wait(context.Background())
+					if err == nil {
+						time.Sleep(5 * time.Millisecond)
+						req.Done()
+						atomic.AddInt32(&completed, 1)
+						sem.Release(1)
+						return
+					}
+					atomic.AddInt32(&retries, 1)
+				}
+			}()
+		}
+	}()
+
+	c.Assert(testutils.WaitUntil(func() bool {
+		return atomic.LoadInt32(&completed) == 1000
+	}, 5*time.Second), check.IsNil)
+
+	log.Infof("%+v", a)
+	log.Infof("Total retries: %v", atomic.LoadInt32(&retries))
+}
+
+func (b *ControllerSuite) TestReservationCancel(c *check.C) {
+	a := NewAPILimiter("foo", APILimiterParameters{
+		RateLimit:           50.0,
+		RateBurst:           10,
+		ParallelRequests:    1,
+		MaxParallelRequests: 1,
+		MaxWaitDuration:     500 * time.Millisecond,
+		Log:                 true,
+	}, nil)
+
+	// Process a request but don't complete it, this will occupy the
+	// parallel request slot
+	req, err := a.Wait(context.Background())
+	c.Assert(err, check.IsNil)
+
+	var completed int32
+
+	// All of these requests must fail due to having to wait too long as
```

Comment:
Looking to clarify, whether these 20 requests fail not because the 500ms wait time is exceeded, but because the one and only parallel slot is taken, and when the slot is taken an attempted request fails immediately.  That's not what I expect.

It is my expectation that the original request will also time out and fail prior to req.Done() being called.  Though that doesn't alter the test, in the sense that the final 10 requests are still expected to succeed, given that the prior 21 (20 in loop and the 1 original) requests have failed.

Reply:
@tgraf Moreover, I'd like to add, that my understanding is that at some point the original request would timeout, and one of the 20 requests in the loop would replace it in the single parallel slot.  Only, to subsequently be cancelled itself for never completing.  I say so, because the original request should time out before the 20 that came after it.

Reply:
Yes, the 20 requests will fail because the only parallel slot available is still held by the request created before the loop. All 20 requests will sit there, waiting for 500ms, and then timeout. The test verifies that the reservation is properly canceled so the next request will immediately succeed.

> It is my expectation that the original request will also time out and fail prior to req.Done() being called.

The first request will succeed as it is already being processed. By the time `Wait()` returns for the first request, the rate limiter has given the OK for the API request to be processed. While it is being processed, the 20 requests get queued up, time out and the the initial request signals completion by calling `req.Done()`. The `MaxWaitDuration` only applies to the waiting until the API request is allowed to be processed.

Reply:
My bad.  I overlooked that the it's "waitDuration", not your traditional request timeout value.